# Promise A+ 规范版本

```js
const PENDING = "pending"; // 进行中
const FULFILLED = "fulfilled"; // 完成
const FAILED = "failed"; // 失败

const isFunction = target => typeof target === "function";

class MyPromise {
  // 当前状态
  state = PENDING;
  // 成功的结果
  value = undefined;
  // 失败的原因
  reason = undefined;
  // 成功回调栈, 先进先执行
  successCallbacks = [];
  // 失败回调栈
  failCallbacks = [];
  constructor(executor) {
    try {
      // executor表示回调函数
      executor(this.resolve, this.reject);
    } catch (e) {
      this.reject(e);
    }
  }

  // 必须使用箭头函数, 因为调用 resolve是直接调用的, 如果不用箭头函数, 绑定的this就会出现错误
  // 箭头函数下 resolve函数内部this始终指向promise实例
  // 否则需要bind
  resolve = (value) => {
    if (this.state === PENDING) {
      // 只有状态为 PENDING才修改状态, 防止重复修改
      this.state = FULFILLED;
      this.value = value;
      while (this.successCallbacks.length) {
        // 执行并出栈
        this.successCallbacks.shift()();
      }
    }
  };

  reject = (reason) => {
    if (this.state === PENDING) {
      this.state = FAILED;
      this.reason = reason;
      while (this.failCallbacks.length) {
        this.failCallbacks.shift()();
      }
    }
  };

  // then为原型链调用, 一般是实例then
  then(successCallback, failCallback) {
    // 很可能两个都不传, 此时需要处理差异
    successCallback = isFunction(successCallback) ? successCallback : v => v;
    failCallback = isFunction(failCallback) ? failCallback : err => new Error(err);
    // then方法返回一个Promise, 以便于链式调用
    let promise2 = new MyPromise((resolve, reject) => {
      const baseQueueMicroTask = (callback, isSuccess = true) => {
        // 创建微任务
        return () => {
          queueMicrotask(() => {
            try {
              // 执行
              let x = callback(isSuccess ? this.value : this.reason);
              // ? 由于目前拿不到promise2, 因此需要让内部代码变成异步的才行
              this.resolvePromise(promise2, x, resolve, reject);
            } catch (e) {
              // then执行过程中报错也要通过reject捕获
              reject(e);
            }
          });
        };
      };
      const successQueueMicroTask = baseQueueMicroTask(successCallback);
      const failQueueMicroTask = baseQueueMicroTask(failCallback, false);
      if (this.state === FULFILLED) {
        successQueueMicroTask();
      } else if (this.state === REJECTED) {
        failQueueMicroTask();
      } else {
        // pending时, 存储成功和失败的回调
        this.successCallbacks.push(successQueueMicroTask);
        this.failCallbacks.push(failQueueMicroTask);
      }
    });
    return promise2;
  }

  // 解析promise
  resolvePromise(promise2, x, resolve, reject) {
    if (promise2 === x) {
      // promise不允许then方法返回自己内部的promise对象, 否则会发生循环引用报错
      return reject(new TypeError("在Promise中出现循环引用!"));
    }
    if (x instanceof MyPromise) {
      // 看then的回调函数执行结果是不是一个promise
      // 如果是 promise实例对象, 需要继续调用then方法查看promise对象的返回结果, 根据结果来决定是调用resolve, 还是reject
      // 这里是 x.then(val => resolve(val), err => reject(err)) 的简写
      x.then(resolve, reject);
    } else {
      resolve(x);
    }
  }

  static all(promiseAll) {
    const res = [];
    let index = 0;
    const addData = (key, value, resolve) => {
      res[key] = value;
      index++;
      if (index === promiseAll.length) {
        // 所有promise都正确执行完成
        resolve(res);
      }
    }
    return new MyPromise((resolve, reject) => {
      for (let i = 0; i < promiseAll.length; i++) {
        const current = promiseAll[i];
        if (current instanceof MyPromise) {
          current.then(value => addData(i, value, resolve), err => {
            reject(err);
          });
        } else {
          addData(i, current, resolve);
        }
      }
    })
  }

  static race(promiseArr) {
    return new MyPromise((resolve, reject) => {
      for (let i = 0; i < promiseArr.length; i++) {
        const current = promiseArr[i];
        if (current instanceof MyPromise) {
          current.then(val => {
            resolve(val);
          }, err => {
            reject(err)
          })
        } else {
          resolve(current);
        }
      }
    });
  }

  // 正确错误都返回, 但是需要带上状态
  static allSettled(promiseArr) {
    const res = [];
    let index = 0;
    const addData = (key, value, resolve, status) => {
      if (status === FULFILLED) {
        res[key] = {
          status: FULFILLED,
          value
        }
      } else {
        res[key] = {
          status: REJECTED,
          reason: value
        }
      }
      index++;
      if (index === promiseArr.length) {
        resolve(res);
      }
    }
    return new MyPromise((resolve) => {
      for (let i = 0; i < promiseArr.length; i++) {
        const current = promiseArr[i];
        if (current instanceof MyPromise) {
          current.then(val => {
            addData(i, val, resolve, FULFILLED);
          }, err => {
            addData(i, err, resolve, REJECTED);
          })
        } else {
          addData(i, current, resolve, FULFILLED);
        }
      }
    })
  }

  static resolve(value) {
    // Promise.resolve 如果内部传入的promise实例, 那么下一次then接收的内容就是这个promise实例执行的结果
    /* 
      Promise.resolve(new Promise((resolve) => {
        console.info("1") // 1
        resolve(100)
      })).then((val) => {
        console.info("then", val) // "then" 100
      })
    */
    if (value instanceof MyPromise) return value;
    return new MyPromise(resolve => resolve(value));
  }

  finally(callback) {
    // 第一都要执行callback
    // 第二返回的是一个promise对象, 后面再then一次, 可以拿到 promise 对象最后一次then的返回结果
    return this.then(val => {
      return MyPromise.resolve(callback()).then(() => val)
    }, err => {
      return MyPromise.resolve(callback()).then(() => { throw err });
    })
  }

  catch (callback) {
    return this.then(undefined, callback);
  }
}
```